test/gtest/core/test_allreduce.cc

```diff
@@ -0,0 +1,180 @@
+                              count * sizeof(typename T::type), UCC_MEMORY_TYPE_HOST,
+                              mem_type);
+            }
+        } else {
+            for (int r = 0; r < ctxs.size(); r++) {
+                dsts[r] = (typename T::type *)(ctxs[r]->args->dst.info.buffer);
+            }
+        }
+        for (int i = 0; i < count; i++) {
+            typename T::type res =
+                    ((typename T::type *)((ctxs[0])->init_buf))[i];
+            for (int r = 1; r < ctxs.size(); r++) {
+                res = T::do_op(res, ((typename T::type *)((ctxs[r])->init_buf))[i]);
+            }
+            for (int r = 0; r < ctxs.size(); r++) {
+                T::assert_equal(res, dsts[r][i]);
+            }
+        }
+        if (UCC_MEMORY_TYPE_HOST != mem_type) {
+            for (int r = 0; r < ctxs.size(); r++) {
+                ucc_mc_free((void*)dsts[r], UCC_MEMORY_TYPE_HOST);
+            }
+        }
+        return;
+    }
+};
+
+TYPED_TEST_CASE(test_allreduce, ReductionTypesOps);
+
+#define TEST_DECLARE(_mem_type, _inplace)                                      \
+{                                                                              \
+    std::array<int,3> counts {1,2,4};                                          \
+    for (int tid = 0; tid < UccJob::nStaticTeams; tid++) {                     \
+        for (int count : counts) {                                             \
+            UccTeam_h team = UccJob::getStaticTeams()[tid];                    \
+            int       size = team->procs.size();                               \
+            this->set_mem_type(_mem_type);                                     \
+            this->set_inplace(_inplace);                                       \
+            UccCollCtxVec ctxs = this->data_init(size, TypeParam::dt, count);  \
+            UccReq    req(team, ctxs);                                         \
+            req.start();                                                       \
+            req.wait();                                                        \
+            this->data_validate(ctxs);                                         \
+            this->data_fini(ctxs);                                             \
+        }                                                                      \
+    }                                                                          \
+}
+
+TYPED_TEST(test_allreduce, single_host) {
+    TEST_DECLARE(UCC_MEMORY_TYPE_HOST, TEST_NO_INPLACE);
+}
+
+TYPED_TEST(test_allreduce, single_host_inplace) {
+    TEST_DECLARE(UCC_MEMORY_TYPE_HOST, TEST_INPLACE);
+}
+
+#ifdef HAVE_CUDA
+TYPED_TEST(test_allreduce, single_cuda) {
+    TEST_DECLARE(UCC_MEMORY_TYPE_CUDA, TEST_NO_INPLACE);
+}
+
+TYPED_TEST(test_allreduce, single_cuda_inplace) {
+    TEST_DECLARE(UCC_MEMORY_TYPE_CUDA, TEST_INPLACE);
+}
+#endif
+
+/* TODO: enable parallel teams once it is supported */
```

Comment:
should already be supported, btw we can enable other multi-team tests as well

Reply:
fixed